### PR TITLE
Request ACCESS_LOCAL_NETWORK when needed

### PR DESCRIPTION
**Pull request by TimoPtr**
## Summary

Add support for the new [ACCESS_LOCAL_NETWORK](https://developer.android.com/reference/android/Manifest.permission#ACCESS_LOCAL_NETWORK) permission. Add to the current flow request to access the local network, the logic is kept dumb without checking the server it is connecting to to keep it simple.

Added to
- onboarding when reaching server discovery (permission with a blank screen behind) if rejected then it redirect to manual server
- WebViewActivity when opening the app it is going to show the error but on top require the permission (it's best effort since we want to drop it)
- FrontendScreen properly ask for permission while the UI is still loading and if rejected proper error screen

## Checklist

- [x] New or updated tests have been added to cover the changes following the testing [guidelines](https://developers.home-assistant.io/docs/android/testing/introduction).
- [x] The code follows the project's [code style](https://developers.home-assistant.io/docs/android/codestyle) and [best_practices](https://developers.home-assistant.io/docs/android/best_practices).
- [x] The changes have been thoroughly tested, and edge cases have been considered.
- [x] Changes are backward compatible whenever feasible. Any breaking changes are documented in the changelog for users and/or in the code for developers depending on the relevance.

## Any other notes

I've ignored a test because Robolectric is not yet ready for API 37. 

Since we are not targetting yet API 37 we might have issues if someone discard the permission from the settings manually but that should probably not happen and we can manually help on discord or github or the user will simply enable it back. 

I've bump the target to 37 and no issue but I don't want yet to make this jump because we need to test more any regression based on the changes made by Google.

### CODE REVIEW

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/haptic/HapticFeedbackPerformerTest.kt
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/haptic/HapticFeedbackPerformerTest.kt:1]: These changes are unrelated

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.ACCESS_WIFI_STATE" />
+    <uses-permission android:name="android.permission.ACCESS_LOCAL_NETWORK" />
```
Comment (jpelgrom) [inline app/src/main/AndroidManifest.xml:9]: Must also be added in `automotive`

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt [hunk 2]
```diff
@@ -533,6 +533,62 @@ class PermissionManagerTest {
 
     // endregion
 
+    // region Local network permission (Android 17+)
+
+    @Nested
+    inner class CheckLocalNetworkPermission {
+
+        @Test
+        fun `Given pre-CINNAMON_BUN device then returns true without checking permission`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt]: We don't use code names anywhere else (other than the constant), update for consistency

```suggestion
        fun `Given pre-API 37 then returns true without checking permission`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt [hunk 3]
```diff
@@ -533,6 +533,62 @@ class PermissionManagerTest {
...
+            val manager = createManager(sdkInt = Build.VERSION_CODES.CINNAMON_BUN - 1)
+            assertTrue(manager.checkLocalNetworkPermission())
+            assertNull(manager.pendingPermissionRequest.value)
+        }
+
+        @Test
+        fun `Given CINNAMON_BUN+ device when permission already granted then returns true`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt]:
```suggestion
        fun `Given API 37 when permission already granted then returns true`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt [hunk 4]
```diff
@@ -533,6 +533,62 @@ class PermissionManagerTest {
...
+            every { permissionChecker.hasPermission(android.Manifest.permission.ACCESS_LOCAL_NETWORK) } returns true
+
+            val manager = createManager(sdkInt = Build.VERSION_CODES.CINNAMON_BUN)
+            assertTrue(manager.checkLocalNetworkPermission())
+            assertNull(manager.pendingPermissionRequest.value)
+        }
+
+        @Test
+        fun `Given CINNAMON_BUN+ device when permission not granted then emits pending request`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt]:
```suggestion
        fun `Given API 37 when permission not granted then emits pending request`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt [hunk 5]
```diff
@@ -533,6 +533,62 @@ class PermissionManagerTest {
...
+            assertEquals(listOf(android.Manifest.permission.ACCESS_LOCAL_NETWORK), pending?.permissions)
+
+            (pending as PermissionRequest.LocalNetwork).onResult(true)
+            advanceUntilIdle()
+            assertTrue(result.await())
+            assertNull(manager.pendingPermissionRequest.value)
+        }
+
+        @Test
+        fun `Given CINNAMON_BUN+ device when user denies then returns false`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/frontend/permissions/PermissionManagerTest.kt]:
```suggestion
        fun `Given API 37 device when user denies then returns false`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/onboarding/serverdiscovery/navigation/ServerDiscoveryNavigationTest.kt [hunk 6]
```diff
@@ -117,6 +120,31 @@ internal class ServerDiscoveryNavigationTest : BaseOnboardingNavigationTest() {
         }
     }
 
+    /**
+     * On Android 17+ (API 37, CINNAMON_BUN), entering the discovery screen requests
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/onboarding/serverdiscovery/navigation/ServerDiscoveryNavigationTest.kt]: One name for the Android version is enough

```suggestion
     * On Android 17+, entering the discovery screen requests
```

File: app/src/test/kotlin/io/homeassistant/companion/android/onboarding/serverdiscovery/ServerDiscoveryViewModelTest.kt [hunk 7]
```diff
@@ -45,6 +45,37 @@ private class ServerDiscoveryViewModelTest {
 
     private fun createViewModel(discoveryMode: ServerDiscoveryMode = ServerDiscoveryMode.NORMAL) {
         viewModel = ServerDiscoveryViewModel(discoveryMode, searcher, serverManager)
+        // Existing tests assert behavior that depends on discovery having started; mirror the
+        // screen's permission-checked callback so they don't all have to opt in individually.
+        viewModel.onLocalNetworkPermissionChecked()
+    }
+
+    @Test
+    fun `Given permission not yet checked then no server discovery is started and NoServerFound is not emitted`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/onboarding/serverdiscovery/ServerDiscoveryViewModelTest.kt]:
```suggestion
    fun `Given local network permission not yet checked then no server discovery is started and NoServerFound is not emitted`() = runTest {
```

File: app/src/test/kotlin/io/homeassistant/companion/android/onboarding/serverdiscovery/ServerDiscoveryViewModelTest.kt [hunk 8]
```diff
@@ -45,6 +45,37 @@ private class ServerDiscoveryViewModelTest {
...
+            assertEquals(Started, awaitItem())
+            advanceTimeBy(TIMEOUT_NO_SERVER_FOUND * 2)
+            runCurrent()
+            // Discovery hasn't started, so the no-server timeout never fires
+            expectNoEvents()
+        }
+    }
+
+    @Test
+    fun `Given permission checked twice then discovery is only started once`() = runTest {
```
Comment (jpelgrom) [inline app/src/test/kotlin/io/homeassistant/companion/android/onboarding/serverdiscovery/ServerDiscoveryViewModelTest.kt]:
```suggestion
    fun `Given local network permission checked twice then discovery is only started once`() = runTest {
```

**Review by jpelgrom — COMMENTED** [review]
Tests in an emulator look promising, I like how minimal the changes required are and how you handled them with the existing setup.

I don't have a good way to test the upgrade path other than manually installing an older version first with a remote URL, then installing this version and changing the URL to local. Of course this change is for behavior which shouldn't happen until targeting even though it does, so I'm fine with keeping this as is for now and testing more when we change the target SDK.

Questions:
- Could you add a note to the changelog like "Added support/fixed local network access on Android 17"? I think it's worth calling out, if only for users wondering why they get the pop-up.
- Worth considering: do we want to opt-in on Android 16 with the 'temporary' permission? That should allow us to discover edge cases sooner.

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/haptic/HapticFeedbackPerformerTest.kt
Comment (TimoPtr) [reply app/src/test/kotlin/io/homeassistant/companion/android/frontend/haptic/HapticFeedbackPerformerTest.kt:1]: It's because of the compile API change that breaks this tests because the function can tak different type of parameters now.

Comment (jpelgrom) [reply app/src/test/kotlin/io/homeassistant/companion/android/frontend/haptic/HapticFeedbackPerformerTest.kt:1]: OK... Do the compile API change separately if this is the only change required? 😇

Comment (TimoPtr) [reply app/src/test/kotlin/io/homeassistant/companion/android/frontend/haptic/HapticFeedbackPerformerTest.kt:1]: It requires a first PR just to fix the test, I found it a bit annoying but I can if you approve this PR tomorrow I'll split it in 2 and do the merge without reviews.

**Comment by jpelgrom** [comment]
> Worth considering: do we want to opt-in on Android 16 with the 'temporary' permission? That should allow us to discover edge cases sooner.

Shortly discussed [on Discord](https://discord.com/channels/330944238910963714/1346948551892009101/1506367366106320991) with @TimoPtr, and I also remembered that it requires [adb opt-in](https://developer.android.com/privacy-and-security/local-network-permission#android-16-guidance), so not that useful for now and we decided to skip.

**Comment by jpelgrom** [comment]
> Could you add a note to the changelog like "Added support/fixed local network access on Android 17"? I think it's worth calling out, if only for users wondering why they get the pop-up.

I think we still need this

**Comment by TimoPtr** [comment]
> > Could you add a note to the changelog like "Added support/fixed local network access on Android 17"? I think it's worth calling out, if only for users wondering why they get the pop-up.
> 
> I think we still need this

I've added it only to the app since it is not relevant today for the rest IMO

**Review by jpelgrom — APPROVED** [review]

File: app/src/test/kotlin/io/homeassistant/companion/android/frontend/haptic/HapticFeedbackPerformerTest.kt
Comment (TimoPtr) [reply app/src/test/kotlin/io/homeassistant/companion/android/frontend/haptic/HapticFeedbackPerformerTest.kt:1]: Done #6871
